File: frontend/src/services/wishlist.ts
import { api } from './api';
import type { Product } from './product';

export interface WishlistItem {
  id: number;
  user_id: number;
  product_id: number;
  created_at: string;
  product?: Product;
}

export interface WishlistDeal {
  product_id: number;
  name: string;
  original_price: number;
  current_price: number;
  discount_percentage: number;
  deal_score: number;
}

export interface WishlistPriceAlert {
  product_id: number;
  name: string;
  old_price: number;
  new_price: number;
  price_drop: number;
  timestamp: string;
}

export interface WishlistAnalytics {
  total_items: number;
  category_distribution: { [category: string]: number };
  average_price: number;
  total_value: number;
  potential_savings: number;
}

export const wishlistService = {
  async getWishlist(): Promise<WishlistItem[]> {
    const res = await api.get<any>('/api/wishlist/');
    return res.data.items || [];
  },

  async addToWishlist(productId: number): Promise<WishlistItem> {
    const res = await api.post<WishlistItem>('/api/wishlist/add', {
      product_id: productId
    });
    return res.data;
  },

  async removeFromWishlist(productId: number): Promise<{ success: boolean; message: string }> {
    const res = await api.delete<{ success: boolean; message: string }>(`/api/wishlist/remove/${productId}`);
    return res.data;
  },

  async clearWishlist(): Promise<{ success: boolean; message: string }> {
    const res = await api.delete<{ success: boolean; message: string }>('/api/wishlist/clear');
    return res.data;
  },

  async moveToCart(productId: number): Promise<{ success: boolean; message: string }> {
    const res = await api.post<{ success: boolean; message: string }>(`/api/wishlist/move-to-cart/${productId}`);
    return res.data;
  },

  async saveForLater(productId: number): Promise<{ success: boolean; message: string }> {
    const res = await api.post<{ success: boolean; message: string }>(`/api/wishlist/save-for-later/${productId}`);
    return res.data;
  },

  async getDealTracking(): Promise<WishlistDeal[]> {
    const res = await api.get<any>('/api/wishlist/deal-tracking');
    return res.data.deals || [];
  },

  async getPriceAlerts(): Promise<WishlistPriceAlert[]> {
    const res = await api.get<any>('/api/wishlist/price-alerts');
    return res.data.alerts || [];
  },

  async getWishlistRecommendations(): Promise<Product[]> {
    const res = await api.get<any>('/api/wishlist/recommendations');
    return res.data.recommendations || [];
  },

  async getWishlistAnalytics(): Promise<WishlistAnalytics> {
    const res = await api.get<WishlistAnalytics>('/api/wishlist/analytics');
    return res.data;
  }
};

File: backend/app/api/wishlist.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel

from app.database.connection import get_db
from app.middleware.auth import get_current_user
from app.services.wishlist_service import WishlistService

router = APIRouter()

wishlist_service = WishlistService()


# ---------- Schemas ----------

class WishlistRequest(BaseModel):
    product_id: int


# ---------- Add Product To Wishlist ----------

@router.post("/add")
def add_to_wishlist(
        request: WishlistRequest,
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.add_to_wishlist(
        db=db,
        user_id=current_user.id,
        product_id=request.product_id
    )


# ---------- Remove Product ----------

@router.delete("/remove/{product_id}")
def remove_from_wishlist(
        product_id: int,
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.remove_from_wishlist(
        db=db,
        user_id=current_user.id,
        product_id=product_id
    )


# ---------- Get Wishlist ----------

@router.get("/")
def get_wishlist(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.get_wishlist(
        db=db,
        user_id=current_user.id
    )


# ---------- Clear Wishlist ----------

@router.delete("/clear")
def clear_wishlist(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.clear_wishlist(
        db=db,
        user_id=current_user.id
    )


# ---------- Move Item To Cart ----------

@router.post("/move-to-cart/{product_id}")
def move_to_cart(
        product_id: int,
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.move_to_cart(
        db=db,
        user_id=current_user.id,
        product_id=product_id
    )


# ---------- Save For Later ----------

@router.post("/save-for-later/{product_id}")
def save_for_later(
        product_id: int,
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.save_for_later(
        db=db,
        user_id=current_user.id,
        product_id=product_id
    )


# ---------- AI Deal Tracking ----------

@router.get("/deal-tracking")
def deal_tracking(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.deal_tracking(
        db=db,
        user_id=current_user.id
    )


# ---------- Price Drop Alerts ----------

@router.get("/price-alerts")
def price_alerts(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.price_alerts(
        db=db,
        user_id=current_user.id
    )


# ---------- Wishlist Recommendations ----------

@router.get("/recommendations")
def wishlist_recommendations(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.wishlist_recommendations(
        db=db,
        user_id=current_user.id
    )


# ---------- Wishlist Analytics ----------

@router.get("/analytics")
def wishlist_analytics(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user)
):

    return wishlist_service.wishlist_analytics(
        db=db,
        user_id=current_user.id
    )


# ---------- Health Check ----------

@router.get("/health")
def health_check():

    return {
        "status": "healthy",
        "service": "wishlist"
    }